## Model Training And Model Selection

In [11]:
import pandas as pd
import numpy as np
# Load datasets
X_train = pd.read_csv(r"D:\ml_udemy\project\datasets\X_train.csv")
X_test = pd.read_csv(r"D:\ml_udemy\project\datasets\X_test.csv")

y_train = pd.read_csv(r"D:\ml_udemy\project\datasets\y_train.csv")
y_test = pd.read_csv(r"D:\ml_udemy\project\datasets\y_test.csv")

print(X_train.shape)
print(X_test.shape)

(12328, 14)
(3083, 14)


In [12]:
## importing all models
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import LinearRegression, Ridge,Lasso
from sklearn.neighbors import KNeighborsRegressor
from xgboost import XGBRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [13]:
##Create a Function to Evaluate Model
def evaluate_model(true, predicted):
    mae = mean_absolute_error(true, predicted)
    mse = mean_squared_error(true, predicted)
    rmse = np.sqrt(mean_squared_error(true, predicted))
    r2_square = r2_score(true, predicted)
    return mae, rmse, r2_square

In [14]:
## Beginning Model Training
models = {
    "Linear Regression": LinearRegression(),
    "Lasso": Lasso(),
    "Ridge": Ridge(),
    "K-Neighbors Regressor": KNeighborsRegressor(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest Regressor": RandomForestRegressor(),
    "Adaboost Regressor":AdaBoostRegressor(),
    "Graident BoostRegressor":GradientBoostingRegressor(),
    "Xgboost Regressor":XGBRegressor()
   
}

for i in range(len(list(models))):
    model = list(models.values())[i]
    model.fit(X_train, y_train) # Train model

    # Make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    
    # Evaluate Train and Test dataset
    model_train_mae , model_train_rmse, model_train_r2 = evaluate_model(y_train, y_train_pred)

    model_test_mae , model_test_rmse, model_test_r2 = evaluate_model(y_test, y_test_pred)

    
    print(list(models.keys())[i])
    
    print('Model performance for Training set')
    print("- Root Mean Squared Error: {:.4f}".format(model_train_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_train_mae))
    print("- R2 Score: {:.4f}".format(model_train_r2))

    print('----------------------------------')
    
    print('Model performance for Test set')
    print("- Root Mean Squared Error: {:.4f}".format(model_test_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_test_mae))
    print("- R2 Score: {:.4f}".format(model_test_r2))
    
    print('='*35)
    print('\n')

Linear Regression
Model performance for Training set
- Root Mean Squared Error: 553855.6665
- Mean Absolute Error: 268101.6071
- R2 Score: 0.6218
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 502543.5930
- Mean Absolute Error: 279618.5794
- R2 Score: 0.6645


Lasso
Model performance for Training set
- Root Mean Squared Error: 553855.6710
- Mean Absolute Error: 268099.2226
- R2 Score: 0.6218
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 502542.6696
- Mean Absolute Error: 279614.7461
- R2 Score: 0.6645


Ridge
Model performance for Training set
- Root Mean Squared Error: 553856.3160
- Mean Absolute Error: 268059.8015
- R2 Score: 0.6218
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 502533.8230
- Mean Absolute Error: 279557.2169
- R2 Score: 0.6645


K-Neighbors Regressor
Model performance for Training set
- Root Mean Squared Error: 325858.7172
- Mean 

d:\ml_udemy\venv\Lib\site-packages\sklearn\base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


Random Forest Regressor
Model performance for Training set
- Root Mean Squared Error: 136815.9732
- Mean Absolute Error: 39983.3585
- R2 Score: 0.9769
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 227842.4276
- Mean Absolute Error: 101431.4414
- R2 Score: 0.9310




d:\ml_udemy\venv\Lib\site-packages\sklearn\utils\validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Adaboost Regressor
Model performance for Training set
- Root Mean Squared Error: 451334.8782
- Mean Absolute Error: 338522.8080
- R2 Score: 0.7488
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 483296.1453
- Mean Absolute Error: 356028.1362
- R2 Score: 0.6897




d:\ml_udemy\venv\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?


Graident BoostRegressor
Model performance for Training set
- Root Mean Squared Error: 204944.5104
- Mean Absolute Error: 111709.5558
- R2 Score: 0.9482
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 254582.2979
- Mean Absolute Error: 126214.9350
- R2 Score: 0.9139


Xgboost Regressor
Model performance for Training set
- Root Mean Squared Error: 85441.7947
- Mean Absolute Error: 61009.3203
- R2 Score: 0.9910
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 246853.8159
- Mean Absolute Error: 99048.5078
- R2 Score: 0.9191




## Model Selection for Hyperparameter Tuning

After evaluating multiple regression models, only the top-performing models were selected for hyperparameter tuning.  
The selection was based on evaluation metrics such as:

- R² Score
- Root Mean Squared Error (RMSE)
- Mean Absolute Error (MAE)
- Generalization performance on test data

### Selected Models

#### 1. Random Forest Regressor
- Achieved the highest Test R² Score.
- Showed strong generalization performance with low error values.
- Reduced overfitting compared to Decision Tree.

#### 2. XGBoost Regressor
- Produced very high prediction accuracy.
- Maintained a good balance between training and testing performance.
- Suitable for handling complex relationships in data.

#### 3. Gradient Boosting Regressor
- Delivered strong predictive performance.
- Performed consistently on unseen test data.
- Effective for improving regression accuracy.

#### 4. K-Neighbors Regressor
- Achieved good test performance with high R² Score.
- Demonstrated strong capability in capturing local data patterns.

---

## Models Not Selected

The following models were not selected for hyperparameter tuning due to comparatively lower performance:

- Linear Regression
- Lasso Regression
- Ridge Regression
- AdaBoost Regressor

Additionally, the Decision Tree model showed signs of overfitting because the training score was extremely high compared to the test score.

---

## Conclusion

Based on the initial model evaluation, the selected models were chosen for hyperparameter tuning to further improve prediction accuracy and optimize model performance.

In [16]:
# Hyperparameter grids for selected models

# Random Forest Regressor
rf_params = {
    "n_estimators": [100, 200, 500],
    "max_depth": [5, 10, 15, None],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4],
    "max_features": ["sqrt", "log2"]
}

# XGBoost Regressor
xgboost_params = {
    "learning_rate": [0.01, 0.05, 0.1],
    "max_depth": [3, 5, 8, 12],
    "n_estimators": [100, 200, 300],
    "subsample": [0.8, 1.0],
    "colsample_bytree": [0.5, 0.8, 1.0]
}

# Gradient Boosting Regressor
gb_params = {
    "learning_rate": [0.01, 0.05, 0.1],
    "n_estimators": [100, 200, 300],
    "max_depth": [3, 5, 8],
    "min_samples_split": [2, 5, 10],
    "subsample": [0.8, 1.0]
}

# K-Neighbors Regressor
knn_params = {
    "n_neighbors": [3, 5, 7, 9, 11],
    "weights": ["uniform", "distance"],
    "algorithm": ["auto", "ball_tree", "kd_tree"],
    "p": [1, 2]
}

In [17]:
# Models list for Hyperparameter tuning

from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.neighbors import KNeighborsRegressor
from xgboost import XGBRegressor

randomcv_models = [

    ("Random Forest", 
     RandomForestRegressor(), 
     rf_params),

    ("XGBoost", 
     XGBRegressor(), 
     xgboost_params),

    ("Gradient Boosting", 
     GradientBoostingRegressor(), 
     gb_params),

    ("KNN Regressor", 
     KNeighborsRegressor(), 
     knn_params)

]

In [18]:
## Hyperparameter Tuning

from sklearn.model_selection import RandomizedSearchCV

randomcv_results = []

# Perform RandomizedSearchCV for each model

for model_name, model, params in randomcv_models:

    print(f"Running RandomizedSearchCV for {model_name}...")

    random_search = RandomizedSearchCV(
        estimator=model,
        param_distributions=params,
        n_iter=10,
        cv=5,
        verbose=2,
        random_state=42,
        n_jobs=-1,
        scoring='r2'
    )

    # Train model
    random_search.fit(X_train, y_train)

    # Store best model and score
    randomcv_results.append({
        "Model": model_name,
        "Best Score": random_search.best_score_,
        "Best Parameters": random_search.best_params_
    })

    print(f"\nBest Parameters for {model_name}:")
    print(random_search.best_params_)

    print(f"Best Cross Validation Score: {random_search.best_score_}")
    print("="*60)

Running RandomizedSearchCV for Random Forest...
Fitting 5 folds for each of 10 candidates, totalling 50 fits


d:\ml_udemy\venv\Lib\site-packages\sklearn\base.py:1389: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)



Best Parameters for Random Forest:
{'n_estimators': 100, 'min_samples_split': 5, 'min_samples_leaf': 1, 'max_features': 'log2', 'max_depth': 15}
Best Cross Validation Score: 0.8684559323269407
Running RandomizedSearchCV for XGBoost...
Fitting 5 folds for each of 10 candidates, totalling 50 fits

Best Parameters for XGBoost:
{'subsample': 1.0, 'n_estimators': 200, 'max_depth': 8, 'learning_rate': 0.05, 'colsample_bytree': 0.8}
Best Cross Validation Score: 0.8454826831817627
Running RandomizedSearchCV for Gradient Boosting...
Fitting 5 folds for each of 10 candidates, totalling 50 fits


d:\ml_udemy\venv\Lib\site-packages\sklearn\ensemble\_gb.py:672: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)  # TODO: Is this still required?



Best Parameters for Gradient Boosting:
{'subsample': 1.0, 'n_estimators': 300, 'min_samples_split': 2, 'max_depth': 5, 'learning_rate': 0.1}
Best Cross Validation Score: 0.8963309127199786
Running RandomizedSearchCV for KNN Regressor...
Fitting 5 folds for each of 10 candidates, totalling 50 fits

Best Parameters for KNN Regressor:
{'weights': 'uniform', 'p': 1, 'n_neighbors': 3, 'algorithm': 'auto'}
Best Cross Validation Score: 0.8740288359870945


## Hyperparameter Tuning Results

Hyperparameter tuning was performed using `RandomizedSearchCV` to optimize the performance of the selected machine learning models.  
The tuning process helped identify the best combination of parameters for each model based on cross-validation performance.

### Best Hyperparameters Obtained

#### 1. Random Forest Regressor
- Best Parameters:
  - n_estimators = 100
  - min_samples_split = 5
  - min_samples_leaf = 1
  - max_features = 'log2'
  - max_depth = 15

- Best Cross Validation Score:
  - 0.8684

---

#### 2. XGBoost Regressor
- Best Parameters:
  - subsample = 1.0
  - n_estimators = 200
  - max_depth = 8
  - learning_rate = 0.05
  - colsample_bytree = 0.8

- Best Cross Validation Score:
  - 0.8454

---

#### 3. Gradient Boosting Regressor
- Best Parameters:
  - subsample = 1.0
  - n_estimators = 300
  - min_samples_split = 2
  - max_depth = 5
  - learning_rate = 0.1

- Best Cross Validation Score:
  - 0.8963

---

#### 4. K-Neighbors Regressor
- Best Parameters:
  - weights = 'uniform'
  - p = 1
  - n_neighbors = 3
  - algorithm = 'auto'

- Best Cross Validation Score:
  - 0.8740

---

## Observation

Among all tuned models, the Gradient Boosting Regressor achieved the highest cross-validation score, indicating strong predictive capability and good generalization performance.

The tuned Random Forest and KNN models also showed strong performance, while XGBoost maintained competitive accuracy after optimization.

---


# Retrain Models with Best Parameters

In [ ]:
# Import models

from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.neighbors import KNeighborsRegressor
from xgboost import XGBRegressor

# Random Forest
rf_best = RandomForestRegressor(
    n_estimators=100,
    min_samples_split=5,
    min_samples_leaf=1,
    max_features='log2',
    max_depth=15,
    random_state=42
)

# XGBoost
xgb_best = XGBRegressor(
    subsample=1.0,
    n_estimators=200,
    max_depth=8,
    learning_rate=0.05,
    colsample_bytree=0.8,
    random_state=42
)

# Gradient Boosting
gb_best = GradientBoostingRegressor(
    subsample=1.0,
    n_estimators=300,
    min_samples_split=2,
    max_depth=5,
    learning_rate=0.1,
    random_state=42
)

# KNN
knn_best = KNeighborsRegressor(
    weights='uniform',
    p=1,
    n_neighbors=3,
    algorithm='auto'
)

# Train Models

In [19]:
## Retraining the models with best parameters

models = {

    "Random Forest Regressor": RandomForestRegressor(
        n_estimators=100,
        min_samples_split=5,
        min_samples_leaf=1,
        max_features='log2',
        max_depth=15,
        random_state=42,
        n_jobs=-1
    ),

    "XGBoost Regressor": XGBRegressor(
        subsample=1.0,
        n_estimators=200,
        max_depth=8,
        learning_rate=0.05,
        colsample_bytree=0.8,
        random_state=42
    ),

    "Gradient Boosting Regressor": GradientBoostingRegressor(
        subsample=1.0,
        n_estimators=300,
        min_samples_split=2,
        max_depth=5,
        learning_rate=0.1,
        random_state=42
    ),

    "KNN Regressor": KNeighborsRegressor(
        weights='uniform',
        p=1,
        n_neighbors=3,
        algorithm='auto'
    )

}


for i in range(len(list(models))):

    model = list(models.values())[i]

    # Train model
    model.fit(X_train, y_train.values.ravel())

    # Make predictions
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)

    # Evaluate Train and Test dataset
    model_train_mae, model_train_rmse, model_train_r2 = evaluate_model(y_train, y_train_pred)

    model_test_mae, model_test_rmse, model_test_r2 = evaluate_model(y_test, y_test_pred)

    print(list(models.keys())[i])

    print('Model performance for Training set')
    print("- Root Mean Squared Error: {:.4f}".format(model_train_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_train_mae))
    print("- R2 Score: {:.4f}".format(model_train_r2))

    print('----------------------------------')

    print('Model performance for Test set')
    print("- Root Mean Squared Error: {:.4f}".format(model_test_rmse))
    print("- Mean Absolute Error: {:.4f}".format(model_test_mae))
    print("- R2 Score: {:.4f}".format(model_test_r2))

    print('='*50)
    print('\n')

Random Forest Regressor
Model performance for Training set
- Root Mean Squared Error: 211080.3800
- Mean Absolute Error: 72643.5303
- R2 Score: 0.9451
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 213594.8865
- Mean Absolute Error: 102131.2056
- R2 Score: 0.9394


XGBoost Regressor
Model performance for Training set
- Root Mean Squared Error: 89983.8720
- Mean Absolute Error: 63243.3438
- R2 Score: 0.9900
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 271212.3976
- Mean Absolute Error: 100205.1797
- R2 Score: 0.9023


Gradient Boosting Regressor
Model performance for Training set
- Root Mean Squared Error: 99818.3930
- Mean Absolute Error: 68815.4381
- R2 Score: 0.9877
----------------------------------
Model performance for Test set
- Root Mean Squared Error: 207412.0349
- Mean Absolute Error: 96482.5079
- R2 Score: 0.9429


KNN Regressor
Model performance for Training set
- Root Mean Squared 

# Our hyperparameter tuning improved the models well.

Now our best model is:

Gradient Boosting

Test R² Score = 0.9429 ✅

Lowest RMSE = 207412

Lowest MAE = 96482


In [23]:
import os
import pickle

# Create models folder if not exists
os.makedirs("models", exist_ok=True)

# Save all tuned models

pickle.dump(models["Random Forest Regressor"], 
            open("models/random_forest_model.pkl", "wb"))

pickle.dump(models["XGBoost Regressor"], 
            open("models/xgboost_model.pkl", "wb"))

pickle.dump(models["Gradient Boosting Regressor"], 
            open("models/gradient_boosting_model.pkl", "wb"))

pickle.dump(models["KNN Regressor"], 
            open("models/knn_model.pkl", "wb"))

print("All models saved successfully")

All models saved successfully
